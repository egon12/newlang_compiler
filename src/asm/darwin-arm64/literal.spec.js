const { Collector } = require('./literal')
const fs = require('fs');
const { asm, Generator } = require('./literal_asm');

describe('literalCollector', () => {

	it('should able to collect global literal', ()=> {
		const collector = new Collector()
		const data = collector.collect(node)

		expect(data).toEqual([
			{ name: 'hello', type: '.ascii', value: 'Hello' },
			{ name: 'main.main.world', type: '.ascii', value: ' world!\n' },
		])
	})

	it('should able to output format in asm', () =>  {
		const collector = new Collector()
		const data = collector.toAsm(node)
		expect(data).toEqual(`
hello: .ascii "Hello"
main.main.world: .ascii " world!
"
`
		)
	})

	it('lets try to compile this tree', () => {
		const generator = new Generator()
		const result = generator.generate(node);
		expect(result).toMatchSnapshot();
		fs.writeFileSync(__dirname + '/literal_asm.spec.s', result);
	})
})

const node = [
	{ 
		type: 'VariableDefinition',
		name: 'hello',
		value: {
			type: 'Literal',
			value: 'Hello'
		}
	},
	{
		type: 'FunctionDefinition',
		namespace: 'main',
		name: 'main',
		parameters: [],
		return: 'int',
		body: [
			{
				type: 'VariableDefinition',
				name: 'world',
				value: {
					type: 'Literal',
					value: ' world!\n'
				},
			},
			{
				type: 'CallStatement',
				value: {
					type: 'CallExpression',
					name: 'syscall3',
					parameters: [
						{ type: 'Number', value: '1' },
						{ type: 'Variable', id: 'world' },
						{ type: 'Number', value: '8' },
						{ type: 'Number', value: '4' },
					]
				}
			},
			{
				type: 'ReturnStatement',
				value: {
					type: 'Number',
					value: '0'
				}
			},
		]
	},
]

const symbolTables = {
	'int': {size: 8, name: 'int'},
	'void': {size: 0, name: 'void'},
	'char': {size: 1, name: 'char'},
	'float': {size: 8, name: 'float'},
	'array_of_char_8': {size: 8, name: 'array', subtype: 'char'},
}
