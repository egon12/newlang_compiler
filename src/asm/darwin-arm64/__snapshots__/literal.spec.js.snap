// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`literalCollector lets try to compile this tree 1`] = `
".global _main
.align 2

_main:
	bl main.main
	mov x16, #1
	svc 0x80
main.main:
	sub sp, sp, #16
	stp x29, x30, [sp, #16]
	add x29, x29, #16
	adrp x0, main.main.world@page
	add x0, x0,main.main.world@pageoff
	str x0, [x29, #-8]
	mov x0, #1
	ldr x1, [x29, #-8]
	mov x2, #8
	mov x3, #4
	mov x16, x3
	svc #0x80
	mov x0, x0
	mov x0, #0
	ldp x29, x30, [sp, #16]
	add sp, sp, #16
	ret
data:

hello: .ascii "Hello"
main.main.world: .ascii " world!
"
"
`;
